### PR TITLE
Introduce `cancel_proposal` to rid us of those pesky proposals

### PR DESCRIPTION
**Pull request by gavofyork**
Also fixes the weight of `propose` and introduces a hard-core `blacklist` option to ensure a proposal hash never returns.

polkadot companion: https://github.com/paritytech/polkadot/pull/1728

TODO: 
- [x] Weights (cc @shawntabrizi )

### CODE REVIEW

File: frame/democracy/src/lib.rs
Comment (joepetrowski) [inline frame/democracy/src/lib.rs:724]:
```suggestion
		/// If called on a queued public or external proposal, then this will result in it being
```

File: frame/democracy/src/lib.rs
Comment (kianenigma) [inline frame/democracy/src/lib.rs:619]: has @jacogr been pinged about this? if not, here's one.

**Review by joepetrowski — COMMENTED** [review]
Breaks `transaction_version` in case that hasn't been updated for the next release.

File: frame/democracy/src/lib.rs
Comment (kianenigma) [inline frame/democracy/src/lib.rs:560]: self/random note that some other modules can generally use this terminology: Some data that is passed in for sanity/weight should be called witness data.

File: frame/democracy/src/lib.rs
Comment (kianenigma) [inline frame/democracy/src/lib.rs:945]: ummm --  didn't we usually add new dispatchables at the bottom to prevent breaking the call indices?

File: frame/democracy/src/lib.rs
Comment (shawntabrizi) [inline frame/democracy/src/lib.rs:197]: In general I would prefer these kinds of consts live as configurable from the pallet's Trait. The feedback we have heard many times from downstream users is the ability to configure their pallet and their weights for their needs and use-cases...

I wont block on it though.

File: frame/democracy/src/lib.rs
Comment (shawntabrizi) [reply frame/democracy/src/lib.rs:945]: +1

**Comment by shawntabrizi** [comment]
I have updated the benchmarks here, and I had to introduce `props_count` to the `blacklist` extrinsic as well.

File: frame/democracy/src/tests.rs
Comment (thiolliere) [inline frame/democracy/src/tests.rs:52]: why redefining MAX_PROPOSAL here instead of using the crate MAX_PROPOSAL ?

File: frame/democracy/src/lib.rs
Comment (thiolliere) [inline frame/democracy/src/lib.rs:619]: why not documenting the new param `prop_count` in the doc ?

db reads in the doc is wrong, it needs one more read of Blacklist, should be better to remove it or update it.

Also it is weird to see `O(1)` while benchmark result in `O(p)`. I see that this `O(p)` is not a cost inside runtime but only inside database (in operation append and maybe also decode_len).

File: frame/democracy/src/lib.rs
Comment (thiolliere) [inline frame/democracy/src/lib.rs:732]: I think documenting prop_count as well would be better

File: frame/democracy/src/lib.rs
Comment (thiolliere) [inline frame/democracy/src/lib.rs:941]:
```suggestion
		/// - `prop_index`: The index of the proposal to cancel.
```

File: frame/democracy/src/lib.rs
Comment (thiolliere) [inline frame/democracy/src/lib.rs:952]: This operation is `O(p)` I think the benchmark is missing this cost.

File: frame/democracy/src/lib.rs
Comment (thiolliere) [inline frame/democracy/src/lib.rs:945]: it also reads PublicProps and potentially writes PublicProps.

File: frame/democracy/src/lib.rs
Comment (thiolliere) [inline frame/democracy/src/lib.rs:749]: this operation also reads and writes PublicProps which is `O(p)`, it seems the benchmark miss it.

**Review by thiolliere — COMMENTED** [review]
some nitpicks and benchmark seems some modification too, put lib logic looks good to me.

File: frame/democracy/src/lib.rs
Comment (thiolliere) [reply frame/democracy/src/lib.rs:749]: ah this one has been fixed

**Comment by gavofyork** [comment]
> Breaks `transaction_version` in case that hasn't been updated for the next release.

**Comment by gavofyork** [comment]
> I have updated the benchmarks here. I had to introduce `props_count` to the `blacklist` extrinsic to take into account its weight relative to the number of proposals.

I didn't do this on purpose.

Firstly `blacklist` is a highly privileged operation and most certainly out of scope for any kind of DDoSing attack. Secondly, it will likely be enacted through some rather long governance process, taking days or even weeks to happen. It is impossible to provide effective witness for this, and requiring it at all just opens up an attack vector whereby an attacker could invalidate the witness immediately prior to enactment through paying for a number of new proposals.

Realistically the weight should either be calculated based only on `max_proposals` (which is, I think, over punitive for the average case), or just accept that on the (very rare) occasion that governance uses this then blocks might *under worst case circumstances) take marginally longer than might otherwise be expected in order to execute.

File: frame/democracy/src/tests.rs
Comment (gavofyork) [reply frame/democracy/src/tests.rs:52]: It is now used.

**Comment by thiolliere** [comment]
if `blacklist(_, _, max_proposals)` weight fits into Operational available weight then we can just do a refund. therefore no punition there.

But I'm ok to have this `out of scope for any kind of DDoSing attack` call having unaccurate weight, just it should be written in the doc I think.

**Comment by gavofyork** [comment]
For now it's safe as `p` is fixed at `MaxProposals` for blacklist. All other grumbles should be addressed.

File: frame/democracy/src/lib.rs
Comment (shawntabrizi) [inline frame/democracy/src/lib.rs:616]: Taking a second look at this function, I think it is fair to call this O(1).

We use `decode_len` in order to figure out the length, which should not need us to decode the whole vector, just the first bytes, then we use `append` to add the new public proposal, which again does not decode/encode the existing proposals, just places it on the end already encoded.

If we look at the benchmark results itself, it is saying that each proposal adds 11 nanoseconds, which is extremely small and within the errors we would expect for the benchmarking pipeline.

I think it would be perfectly safe to keep treating this as O(1) and remove the `prop_count`.

What do you think?

File: frame/democracy/src/lib.rs
Comment (shawntabrizi) [inline frame/democracy/src/lib.rs:1182]:
```suggestion
		/// The dispatch origin of this call must be _CancelProposalOrigin_.
```

File: frame/democracy/src/lib.rs
Comment (shawntabrizi) [inline frame/democracy/src/lib.rs:1170]: This is not allowed to trigger an error. You will need to explicitly handle the result.

**Review by shawntabrizi — CHANGES_REQUESTED** [review]
Other than my comments, this looks good.

**Comment by gavofyork** [comment]
@kianenigma @thiolliere @shawntabrizi ready for final review...

File: frame/democracy/src/lib.rs
Comment (thiolliere) [inline frame/democracy/src/lib.rs:607]: This is no longer an argument.
```suggestion
```

File: frame/democracy/src/lib.rs
Comment (thiolliere) [inline frame/democracy/src/lib.rs:1185]: just an after thought: couldn't we have used `Props::decode_len()` here ?
decode_len only do a `sp_io::storage::read(key, &mut array_of_length_5, 0)?;` and here the key is unique so seems fine.

(and same for blacklist)

**Review by thiolliere — APPROVED** [review]
looks good to me

File: frame/democracy/src/lib.rs
Comment (gavofyork) [reply frame/democracy/src/lib.rs:1185]: prefer to avoid any databse access in the weight calculation as it can be spammed in the txq.

**Comment by gavofyork** [comment]
@kianenigma @shawntabrizi would be good to get a final review.

**Comment by shawntabrizi** [comment]
`on_initialize` weights look borked when running benchmarks again:

```
// before
		(70826000 as Weight)
			.saturating_add((10716000 as Weight).saturating_mul(r as Weight))
			.saturating_add(T::DbWeight::get().reads(6 as Weight))
			.saturating_add(T::DbWeight::get().reads((2 as Weight).saturating_mul(r as Weight)))
			.saturating_add(T::DbWeight::get().writes(5 as Weight))
//after
		(12_357_000 as Weight)
			.saturating_add((9_372_000 as Weight).saturating_mul(r as Weight))
			.saturating_add(T::DbWeight::get().reads(5 as Weight))
			.saturating_add(T::DbWeight::get().reads((1 as Weight).saturating_mul(r as Weight)))
```

Missing reads and writes. I am investigating

**Review by shawntabrizi — APPROVED** [review]
I pushed the latest Weights to the PR. Note that the only relevant changes was an additional read to the `propose` weight, caused by checking the blacklist, and a reduction in the `on_initialize_base` weight due to bad benchmarking code.

To clarify the intention, `on_initialize_base` is supposed to be the least amount of computation needed by the `on_initialize` code, so in this case (5 + R) Reads.

If the on_initialize code actually triggers any of the more complex paths, like `launch_next` or `bake_referendum`, we will assume that the computation should take a full block weight. This is hardcoded into the `on_initialize` logic.

Tracking Issue: https://github.com/paritytech/substrate/issues/7209

**Comment by gavofyork** [comment]
bot merge

**Pull request merged**
